controller/setting_controller.go

```diff
@@ -322,27 +324,62 @@ func (sc *SettingController) updateTaintToleration() error {
 	}
 
 	for _, dp := range deploymentList {
-		if util.AreIdenticalTolerations(util.TolerationListToMap(dp.Spec.Template.Spec.Tolerations), newTolerations) {
+		lastAppliedTolerationsList, err := getLastAppliedTolerationsList(dp)
+		if err != nil {
+			return err
+		}
+		if reflect.DeepEqual(util.TolerationListToMap(lastAppliedTolerationsList), newTolerationsMap) {
 			continue
 		}
-		dp.Spec.Template.Spec.Tolerations = getFinalTolerations(util.TolerationListToMap(dp.Spec.Template.Spec.Tolerations), newTolerations)
+
+		dp.Spec.Template.Spec.Tolerations = getFinalTolerations(util.TolerationListToMap(dp.Spec.Template.Spec.Tolerations),
```

Comment:
How about using a single function to update tolerations and the related annotations for all Longhorn workloads? There are too many details of operating tolerations here, which is not straightforward.

Similar below.

Reply:
Do you mean that we use a separate function to update both tolerations and annotations?
If so, yeah, I agree!

Reply:
Fixed. Thanks